chain/client/src/stateless_validation/chunk_endorsement_tracker.rs

```diff
@@ -112,65 +106,104 @@ impl ChunkEndorsementTracker {
     /// Otherwise, we store the endorsement in a separate cache of endorsements to be processed when the chunk is ready.
     fn process_chunk_endorsement_impl(
-        &self,
+        &mut self,
         endorsement: ChunkEndorsement,
         chunk_header: Option<&ShardChunkHeader>,
     ) -> Result<(), Error> {
         let chunk_hash = endorsement.chunk_hash();
         let account_id = &endorsement.account_id;
 
-        let endorsement_cache = if chunk_header.is_some() {
-            &self.chunk_endorsements
-        } else {
-            &self.pending_chunk_endorsements
-        };
+        let existing_entry = self.chunk_endorsements.peek(chunk_hash);
 
         // If we have already processed this chunk endorsement, return early.
-        if endorsement_cache
-            .get(chunk_hash)
-            .is_some_and(|existing_endorsements| existing_endorsements.contains_key(account_id))
-        {
+        if existing_entry.is_some_and(|(_, existing_endorsements)| {
+            existing_endorsements.contains_key(account_id)
+        }) {
             tracing::debug!(target: "client", ?endorsement, "Already received chunk endorsement.");
             return Ok(());
         }
 
-        if let Some(chunk_header) = chunk_header {
-            if !self.epoch_manager.verify_chunk_endorsement(&chunk_header, &endorsement)? {
-                tracing::error!(target: "client", ?endorsement, "Invalid chunk endorsement.");
-                return Err(Error::InvalidChunkEndorsement);
-            }
-        }
-
         // If we are the current block producer, we store the chunk endorsement for each chunk which
         // would later be used during block production to check whether to include the chunk or not.
         // TODO(stateless_validation): It's possible for a malicious validator to send endorsements
         // for 100 unique chunks thus pushing out current valid endorsements from our cache.
         // Maybe add check to ensure we don't accept endorsements from chunks already included in some block?
         // Maybe add check to ensure we don't accept endorsements from chunks that have too old height_created?
         tracing::debug!(target: "client", ?endorsement, "Received and saved chunk endorsement.");
-        let mut guard = endorsement_cache.lock();
-        guard.get_or_insert(chunk_hash.clone(), || HashMap::new());
-        let chunk_endorsements = guard.get_mut(chunk_hash).unwrap();
-        chunk_endorsements.insert(account_id.clone(), endorsement);
+
+        // The header might be available in the endorsement cache, even if it isn't provided.
+        // In such case it should be treated as a non-pending endorsement.
+        let header = chunk_header.or_else(|| existing_entry.map(|(header, _)| header));
+
+        if let Some(chunk_header) = header {
+            if !self.epoch_manager.verify_chunk_endorsement(&chunk_header, &endorsement)? {
```

Comment:
I haven't read the logic in detail; just a question about this - can this be moved to outside the mutex? I'm not sure how expensive this is.